File: 6.x/c/example3.c
/* Example 3 - Using our own performance loop
 * Author: Steven Yi <stevenyi@gmail.com>
 * 2013.10.28
 *
 * In this example, we use a while loop to perform Csound one audio block at a time.
 * This technique is important to know as it will allow us to do further processing
 * safely at block boundaries.  We will explore the technique further in later examples. 
 */


#include <stdio.h>
#include <csound.h>

/* Defining our Csound ORC code within a multiline String */
const char* orc = "sr=44100\n"
  "ksmps=32\n"
  "nchnls=2\n"
  "0dbfs=1\n\n"
  "instr 1\n"
  "aout vco2 0.5, 440\n"
  "outs aout, aout\n"
  "endin";

/*Defining our Csound SCO code */
const char* sco = "i1 0 1";

int main(int arg, char** argv) {

  csoundInitialize(CSOUNDINIT_NO_ATEXIT);

  CSOUND* csound = csoundCreate(NULL);

  /* Using SetOption() to configure Csound 
  Note: use only one commandline flag at a time */
  csoundSetOption(csound, "-odac");

  /* Compile the Csound Orchestra string */
  csoundCompileOrc(csound, orc);

  /* Compile the Csound SCO String */
  csoundReadScore(csound, (char*)sco);

  /* When compiling from strings, this call is necessary 
   * before doing any performing */
  csoundStart(csound);

  /* The following is our main performance loop. We will perform one 
   * block of sound at a time and continue to do so while it returns 0, 
   * which signifies to keep processing.  We will explore this loop 
   * technique in further examples. 
   */

  while (csoundPerformKsmps(csound) == 0) {
    /* pass for now */
  }

  csoundStop(csound);

  return 0;
}

File: 6.x/csharp/Example3.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using csound6netlib;  //Exposes DotNet to Csound 6 Bridge's classes from csound6netlib.dll

namespace csoundAPI_examples
{
    public partial class CsoundAPI_Examples
    {
       /* Example 3 - Using our own performance loop
        * 
        * In this example, we use a while loop to perform Csound one audio block at a time.
        * This technique is important to know as it will allow us to do further processing
        * safely at block boundaries.  We will explore the technique further in later examples.
        * 
        * The C# version includes checking the status code which csound returns after key processing.
        * These codes are encapsulated in the CsoundStatus enumeration.
        * This demonstrates the traditional way to respond to csound compiling and performing
        */
        public void Example3()
        {
            const string sco = "i1 0 1\n";
            using (var c = new Csound6Net())
            {
                //You can also set the output file or device using SetOutputFileName method in the API
                // SetOutputFileName(string path, SoundFileType type, SampleFormat format)
                // or its convenience method for real time output SetOutputDac(int dacNbr)
                // instead of via command line arguments fed to SetOption(string option) one at a time
                c.SetOutputDac(0);

                var status = c.CompileOrc(orc);      // Compile Orchestra from String
                if (status == CsoundStatus.Success)  // Classic csound practice tests compile results before proceeding 
                {
                    status = c.ReadScore(sco);       // Read in Score from String variable
                    if (status == CsoundStatus.Success)
                    {
                        c.Start(); // When compiling from strings, this call needed before performing

                        /* The following is our main performance loop.
                         * We will perform one block of sound at a time and continue to do so
                         * while it returns false, which tells us to keep processing.
                         * We will explore this loop technique in further examples.
                         */
                        while (!c.PerformKsmps());

                        c.Stop();
                    }
                }
            }
        }


    }
}
